File: analysis/anagram.py
# Determine whether s1 is an anagram of s2
# alg: 
#   1. Check if letter in s1 is in s2
#   2. if not, return false
#   3. else, checkoff letter from s2
#   4. return tru
def is_anagram_checking_off(s1, s2):
  if len(s1) != len(s2):
    return False

  checked_off = list(s2)

  for char1 in s1:
    for i, char2 in enumerate(checked_off):
      if char1 == char2:
        checked_off[i] = None
        break
    else: # no break
      return False
  return True
      
print is_anagram_checking_off('hello', 'ellho')
print is_anagram_checking_off('hello', 'elvho')

File: analysis/anagram.js
/**
 * 1. Check if letter in s1 is in s2
 * 2. if not, return false
 * 3. else, checkoff letter from s2
 * 4. return true
 * @param {string} s1 
 * @param {string} s2 
 */
const isAnagram = (s1, s2) => {
  isCheckedOff = {};
  const { length } = s1;
  if (length !== s2.length) {
    return false;
  }

  for (let i = 0; i < length; i++) {
    const c1 = s1[i];
    let isLetterFound = false
    for (let j = 0; j < length; j++) {
      if (!isCheckedOff[j] && s2[j] === c1) {
        isCheckedOff[j] = true;
        isLetterFound = true;
        break;
      }
    }
    if (!isLetterFound) {
      return false;
    }
  }
  
  return true;
};

console.log(isAnagram('elo', 'leo'));
console.log(isAnagram('elo', 'ley'));

